File: plugins/ethics_ai/philosophy_loader.cpp
/*
╔═════════════════════════════════════════════════════════════════════╗
║ ThemisDB - Hybrid Database System                                   ║
╠═════════════════════════════════════════════════════════════════════╣
  File:            philosophy_loader.cpp                              ║
  Version:         0.0.39                                             ║
  Last Modified:   2026-04-13 20:28:40                                ║
  Author:          unknown                                            ║
╠═════════════════════════════════════════════════════════════════════╣
  Quality Metrics:                                                    ║
    • Maturity Level:  🟢 PRODUCTION-READY                             ║
    • Quality Score:   100.0/100                                      ║
    • Total Lines:     181                                            ║
    • Open Issues:     TODOs: 0, Stubs: 0                             ║
╠═════════════════════════════════════════════════════════════════════╣
  Revision History:                                                   ║
    • 2a1fb04231  2026-03-03  Merge branch 'develop' into copilot/audit-src-module-docu... ║
╠═════════════════════════════════════════════════════════════════════╣
  Status: ✅ Production Ready                                          ║
╚═════════════════════════════════════════════════════════════════════╝
 */

#include "philosophy_loader.h"
#include <fstream>
#include <sstream>
#include <filesystem>

#ifdef HAVE_YAML_CPP
#include <yaml-cpp/yaml.h>
#endif

namespace themis {
namespace plugins {
namespace ethics {

std::variant<size_t, Status> PhilosophyLoader::loadFromDirectory(const std::string& directory) {
    namespace fs = std::filesystem;
    
    if (!fs::exists(directory)) {
        return Status::Error("Directory does not exist: " + directory);
    }
    
    if (!fs::is_directory(directory)) {
        return Status::Error("Path is not a directory: " + directory);
    }
    
    size_t count = 0;
    for (const auto& entry : fs::directory_iterator(directory)) {
        if (entry.is_regular_file()) {
            auto path = entry.path();
            auto ext = path.extension().string();
            
            if (ext == ".yaml" || ext == ".yml") {
                auto status = loadFromFile(path.string());
                if (status.isOK()) {
                    count++;
                }
            }
        }
    }
    
    return count;
}

Status PhilosophyLoader::loadFromFile(const std::string& filepath) {
#ifdef HAVE_YAML_CPP
    try {
        YAML::Node config = YAML::LoadFile(filepath);
        
        PhilosophyProfile profile;
        profile.school_id = config["school_id"].as<std::string>("");
        profile.name = config["name"].as<std::string>("");
        
        if (profile.school_id.empty()) {
            // Use filename without extension as school_id
            namespace fs = std::filesystem;
            profile.school_id = fs::path(filepath).stem().string();
        }
        
        // Load main theses
        if (config["main_theses"]) {
            for (const auto& thesis : config["main_theses"]) {
                profile.main_theses.push_back(thesis.as<std::string>());
            }
        }
        
        // Load secondary theses
        if (config["secondary_theses"]) {
            for (const auto& thesis : config["secondary_theses"]) {
                profile.secondary_theses.push_back(thesis.as<std::string>());
            }
        }
        
        // Load decision framework
        if (config["decision_framework"]) {
            for (const auto& kv : config["decision_framework"]) {
                profile.decision_framework[kv.first.as<std::string>()] = 
                    kv.second.as<std::string>();
            }
        }
        
        // Load strengths
        if (config["strengths"]) {
            for (const auto& strength : config["strengths"]) {
                profile.strengths.push_back(strength.as<std::string>());
            }
        }
        
        // Load weaknesses
        if (config["weaknesses"]) {
            for (const auto& weakness : config["weaknesses"]) {
                profile.weaknesses.push_back(weakness.as<std::string>());
            }
        }
        
        // Load internal debate
        if (config["internal_debate"]) {
            for (const auto& kv : config["internal_debate"]) {
                profile.internal_debate[kv.first.as<std::string>()] = 
                    kv.second.as<std::string>();
            }
        }
        
        // Load philosophical positioning
        if (config["philosophical_positioning"]) {
            for (const auto& kv : config["philosophical_positioning"]) {
                profile.philosophical_positioning[kv.first.as<std::string>()] = 
                    kv.second.as<std::string>();
            }
        }
        
        profiles_[profile.school_id] = profile;
        return Status::OK();
        
    } catch (const YAML::Exception& e) {
        return Status::Error("YAML parsing error: " + std::string(e.what()));
    } catch (const std::exception& e) {
        return Status::Error("Error loading file: " + std::string(e.what()));
    }
#else
    return Status::Error("YAML support not enabled (yaml-cpp not found)");
#endif
}

std::variant<PhilosophyProfile, Status> PhilosophyLoader::getProfile(
    const std::string& school_id) const {
    
    auto it = profiles_.find(school_id);
    if (it == profiles_.end()) {
        return Status::Error("Philosophy profile not found: " + school_id);
    }
    
    return it->second;
}

bool PhilosophyLoader::hasProfile(const std::string& school_id) const {
    return profiles_.find(school_id) != profiles_.end();
}

std::vector<std::string> PhilosophyLoader::getSchoolIds() const {
    std::vector<std::string> ids;
    ids.reserve(profiles_.size());
    
    for (const auto& kv : profiles_) {
        ids.push_back(kv.first);
    }
    
    return ids;
}

void PhilosophyLoader::clear() {
    profiles_.clear();
}

std::map<std::string, PhilosophyProfile> PhilosophyLoader::getAllProfiles() const {
    return profiles_;
}

} // namespace ethics
} // namespace plugins
} // namespace themis

File: examples/24_moral_philosophy_debates/philosophy_loader.py
"""
╔═════════════════════════════════════════════════════════════════════╗
║ ThemisDB - Hybrid Database System                                   ║
╠═════════════════════════════════════════════════════════════════════╣
  File:            philosophy_loader.py                               ║
  Version:         0.0.44                                             ║
  Last Modified:   2026-04-15 05:32:38                                ║
  Author:          unknown                                            ║
╠═════════════════════════════════════════════════════════════════════╣
  Quality Metrics:                                                    ║
    • Maturity Level:  🟢 PRODUCTION-READY                             ║
    • Quality Score:   100.0/100                                      ║
    • Total Lines:     284                                            ║
    • Open Issues:     TODOs: 0, Stubs: 0                             ║
╠═════════════════════════════════════════════════════════════════════╣
  Revision History:                                                   ║
    • 2a1fb04231  2026-03-03  Merge branch 'develop' into copilot/audit-src-module-docu... ║
╠═════════════════════════════════════════════════════════════════════╣
  Status: ✅ Production Ready                                          ║
╚═════════════════════════════════════════════════════════════════════╝
"""

"""
Philosophy Profile Loader

Lädt Philosophy-Profile aus YAML-Dateien im philosophies/ Verzeichnis.
Unterstützt dynamisches Laden und Caching.
"""

import os
import yaml
from pathlib import Path
from typing import Dict, Optional, List
from dataclasses import dataclass
from models import PhilosophySchool, PhilosophyProfile


class PhilosophyLoader:
    """
    Lädt und verwaltet Philosophy-Profile aus YAML-Dateien.
    """
    
    def __init__(self, philosophies_dir: str = "philosophies"):
        """
        Initialize the philosophy loader.
        
        Args:
            philosophies_dir: Verzeichnis mit YAML-Dateien
        """
        self.philosophies_dir = Path(philosophies_dir)
        self._cache: Dict[PhilosophySchool, PhilosophyProfile] = {}
        self._yaml_files: Dict[PhilosophySchool, Path] = {}
        self._discover_yaml_files()
    
    def _discover_yaml_files(self):
        """Entdeckt alle YAML-Dateien im Verzeichnis."""
        if not self.philosophies_dir.exists():
            print(f"Warning: Directory {self.philosophies_dir} does not exist")
            return
        
        # Mapping von school-Namen zu YAML-Dateien
        for yaml_file in self.philosophies_dir.glob("*.yaml"):
            try:
                with open(yaml_file, 'r', encoding='utf-8') as f:
                    data = yaml.safe_load(f)
                    if data and 'school' in data:
                        school_str = data['school']
                        try:
                            school = PhilosophySchool(school_str)
                            self._yaml_files[school] = yaml_file
                        except ValueError:
                            print(f"Warning: Unknown school '{school_str}' in {yaml_file}")
            except Exception as e:
                print(f"Error reading {yaml_file}: {e}")
    
    def load_profile(self, school: PhilosophySchool) -> Optional[PhilosophyProfile]:
        """
        Lädt ein Philosophy-Profile aus YAML-Datei.
        
        Args:
            school: Die zu ladende Philosophie
        
        Returns:
            PhilosophyProfile oder None wenn nicht gefunden
        """
        # Aus Cache zurückgeben wenn vorhanden
        if school in self._cache:
            return self._cache[school]
        
        # YAML-Datei laden
        if school not in self._yaml_files:
            return None
        
        yaml_file = self._yaml_files[school]
        try:
            with open(yaml_file, 'r', encoding='utf-8') as f:
                data = yaml.safe_load(f)
            
            # Konvertiere YAML zu PhilosophyProfile
            profile = self._yaml_to_profile(data, school)
            
            # In Cache speichern
            self._cache[school] = profile
            
            return profile
        
        except Exception as e:
            print(f"Error loading profile for {school.value}: {e}")
            return None
    
    def _yaml_to_profile(self, data: dict, school: PhilosophySchool) -> PhilosophyProfile:
        """
        Konvertiert YAML-Daten zu PhilosophyProfile.
        
        Args:
            data: YAML-Daten als Dictionary
            school: PhilosophySchool Enum
        
        Returns:
            PhilosophyProfile Objekt
        """
        # Extrahiere core_principles aus YAML
        core_principles = data.get('core_principles', [])
        if isinstance(core_principles, list):
            # Liste von Strings
            core_principles_list = core_principles
        else:
            # Fallback zu leerer Liste
            core_principles_list = []
        
        # Erstelle description aus main_theses wenn nötig
        description = data.get('description', '')
        if not description and 'main_theses' in data:
            # Baue description aus ersten Hauptthesen
            main_theses = data['main_theses']
            if main_theses:
                first_thesis = list(main_theses.values())[0]
                description = first_thesis.get('description', '')
        
        # Erstelle PhilosophyProfile
        profile = PhilosophyProfile(
            school=school,
            name=data.get('name', school.value),
            philosopher_name=data.get('philosopher_name', ''),
            description=description.strip() if description else '',
            core_principles=core_principles_list,
            decision_framework=data.get('decision_framework', ''),
            example_application=data.get('example_application', '')
        )
        
        return profile
    
    def load_all_profiles(self) -> Dict[PhilosophySchool, PhilosophyProfile]:
        """
        Lädt alle verfügbaren Philosophy-Profile.
        
        Returns:
            Dictionary mit allen Profilen
        """
        profiles = {}
        for school in self._yaml_files.keys():
            profile = self.load_profile(school)
            if profile:
                profiles[school] = profile
        return profiles
    
    def get_available_schools(self) -> List[PhilosophySchool]:
        """
        Gibt Liste aller verfügbaren Schools zurück.
        
        Returns:
            Liste von PhilosophySchool Enums
        """
        return list(self._yaml_files.keys())
    
    def reload_profile(self, school: PhilosophySchool) -> Optional[PhilosophyProfile]:
        """
        Lädt ein Profile neu (ohne Cache).
        
        Args:
            school: Die neu zu ladende Philosophie
        
        Returns:
            PhilosophyProfile oder None
        """
        # Entferne aus Cache
        if school in self._cache:
            del self._cache[school]
        
        # Neu laden
        return self.load_profile(school)
    
    def get_profile_metadata(self, school: PhilosophySchool) -> Optional[dict]:
        """
        Gibt Metadaten eines Profiles zurück ohne vollständiges Laden.
        
        Args:
            school: Die Philosophie
        
        Returns:
            Dictionary mit Metadaten
        """
        if school not in self._yaml_files:
            return None
        
        yaml_file = self._yaml_files[school]
        try:
            with open(yaml_file, 'r', encoding='utf-8') as f:
                data = yaml.safe_load(f)
            
            return {
                'name': data.get('name', ''),
                'philosopher_name': data.get('philosopher_name', ''),
                'philosopher_life': data.get('philosopher_life', ''),
                'nationality': data.get('nationality', ''),
                'has_main_theses': 'main_theses' in data,
                'has_secondary_theses': 'secondary_theses' in data,
                'key_works_count': len(data.get('key_works', [])),
                'file': str(yaml_file)
            }
        except Exception as e:
            print(f"Error reading metadata for {school.value}: {e}")
            return None


# Globaler Loader (Singleton)
_global_loader: Optional[PhilosophyLoader] = None


def get_philosophy_loader(philosophies_dir: str = "philosophies") -> PhilosophyLoader:
    """
    Gibt den globalen Philosophy-Loader zurück (Singleton).
    
    Args:
        philosophies_dir: Verzeichnis mit YAML-Dateien
    
    Returns:
        PhilosophyLoader Instanz
    """
    global _global_loader
    if _global_loader is None:
        _global_loader = PhilosophyLoader(philosophies_dir)
    return _global_loader


def load_philosophy_profiles() -> Dict[PhilosophySchool, PhilosophyProfile]:
    """
    Convenience-Funktion zum Laden aller Profile.
    
    Returns:
        Dictionary mit allen verfügbaren Profilen
    """
    loader = get_philosophy_loader()
    return loader.load_all_profiles()


# Beispiel-Verwendung
if __name__ == "__main__":
    # Lade alle Profile
    loader = get_philosophy_loader()
    
    print("Verfügbare Philosophien:")
    for school in loader.get_available_schools():
        metadata = loader.get_profile_metadata(school)
        if metadata:
            print(f"\n{metadata['name']} ({metadata['philosopher_name']})")
            print(f"  Leben: {metadata['philosopher_life']}")
            print(f"  Datei: {metadata['file']}")
            print(f"  Hauptthesen: {'Ja' if metadata['has_main_theses'] else 'Nein'}")
            print(f"  Werke: {metadata['key_works_count']}")
    
    print("\n" + "="*60)
    print("Lade alle Profile...")
    profiles = loader.load_all_profiles()
    print(f"✓ {len(profiles)} Profile geladen")
    
    # Beispiel: Kant laden
    if PhilosophySchool.KANT in profiles:
        kant = profiles[PhilosophySchool.KANT]
        print(f"\nBeispiel - {kant.name}:")
        print(f"Philosoph: {kant.philosopher_name}")
        print(f"Kernprinzipien: {len(kant.core_principles)}")
        print(f"Erstes Prinzip: {kant.core_principles[0] if kant.core_principles else 'N/A'}")
